### ISSUE TITLE
Experiment -- Reset player when ad is detected

### ISSUE THREAD

**Issue by younesaassila** [younesaassila/ttv-lol-pro#32]
Implement FFZ Reset Player feature and experiment with having it trigger when an ad is playing.

**Comment by thinkpad4** [comment]
Would it be possible to make that a toggle option?  Thanks

**Comment by younesaassila** [comment]
> Would it be possible to make that a toggle option?  Thanks

Of course

**Comment by younesaassila** [comment]
Added `help wanted` label since implementation seems to be **A PAIN**. It looks like FFZ is injecting a script to get access to React's Fiber tree and to access player functions? Arghhh

**Comment by younesaassila** [comment]
🚧 Managed to retrieve the Media Player Instance, still need to do all the TODOs in the code 🤡

**Comment by younesaassila** [comment]
I'm trying to detect if a midroll ad is about to play the same way [VideoAdBlockForTwitch](https://github.com/cleanlock/VideoAdBlockForTwitch) is doing it. Unfortunately, this requires injecting a script and modifying the default `Worker` instance. It's a mess...

For now, I'm planning to use `filterResponseData` (available on Firefox only) to inspect `video-weaver` responses and double-clicking the "Reset Player" button of FFZ (if FFZ is installed) if an ad is detected.

Perhaps detecting the ad could simply be done using an Observer on some DOM element, but I can't seem to be able to get an ad to play to check how it could be done. Adding back the `help wanted` tag, anyone with a solution, feel free to help!

**Comment by juliocoliveira** [comment]
The full css path for a ad counter is: html.tw-root--hover.js-focus-visible.tw-root--theme-dark body.logged-in.dark-theme.seventv-no-paints.bttv-hide-player-extensions.bttv-rs-theme-dark.bttv-hide-player-cursor div#root.root div.Layout-sc-1xcs6mc-0.hRpbps div.Layout-sc-1xcs6mc-0.kBprba div.Layout-sc-1xcs6mc-0.dajtya main.InjectLayout-sc-1i43xsx-0.iqhyKW.twilight-main div.root-scrollable.scrollable-area.scrollable-area--suppress-scroll-x div.simplebar-scroll-content div.simplebar-content.root-scrollable__content div.Layout-sc-1xcs6mc-0.hUemkh.root-scrollable__wrapper div.InjectLayout-sc-1i43xsx-0.persistent-player div.ScAspectRatio-sc-18km980-1.hTTohL.tw-aspect div.Layout-sc-1xcs6mc-0.video-player div.Layout-sc-1xcs6mc-0.kUDtlR.video-player__container.video-player__container--resize-calc div.Layout-sc-1xcs6mc-0.video-ref div.Layout-sc-1xcs6mc-0.video-player__default-player.video-player__inactive div.Layout-sc-1xcs6mc-0.kfOuIF.video-player__overlay div.Layout-sc-1xcs6mc-0.fYzbEF div.Layout-sc-1xcs6mc-0.kDYniE You can use attribute check for div data-a-target="video-ad-countdown"
![image](https://user-images.githubusercontent.com/6606725/209483515-c097138d-8228-444c-b29b-4fa507125105.png)

**Comment by younesaassila** [comment]
Okay so I've fully implemented the feature for Firefox, should it be enabled by default? (The code still needs a bit of polishing perhaps)

**Comment by juliocoliveira** [comment]
Yes it should but you should add a option to enable or not.
It will be very annoying some times i have to reload 3,4 times in a row to get rid of the ads. But is better than seeing the ads.

**Comment by younesaassila** [comment]
I've released a [beta version of version 1.6.0 which includes this feature](https://github.com/younesaassila/ttv-lol-pro/releases/tag/v1.6.0-beta.0). I'm looking for feedback on how well the feature works/doesn't work.

**The feature is Firefox only!** Also, this beta has auto-update turned off and must be installed as a temporary extension since it is not signed. I'd recommend installing it in another browser like Firefox Developer Edition. If you want to stick to your regular installation of Firefox, don't forget to disable the stable version of TTV LOL PRO when testing the beta!

**Comment by juliocoliveira** [comment]
I tested on Firefox developer edition, and so far no ads, YAY!, Until Twitch circumvents this too ¯\_(ツ)_/¯

**Comment by jakethewake** [comment]
Seems to work really well, i asked my friend to run ads the stream refreshed twice and i got no ads. 👍

**Comment by younesaassila** [comment]
Glad to know resetting the player seems to do the trick for midrolls! 🚀

Version 1.6.0 is pretty much ready to ship (I think I'll delay #17 once again, even though it was planned for this version)

**Comment by thinkpad4** [comment]
Ready to ship on Chrome and Firefox?

**Comment by younesaassila** [comment]
> Ready to ship on Chrome and Firefox?

"Reset player on midroll" will be Firefox only for now. We'll have to release some big Chrome update in the future with Reset feature, Proxy country feature, and auto-updates.

**Comment by juliocoliveira** [comment]
When can we expect to see the 1.6.0 update on the release/stable/main channel/branch?

**Comment by younesaassila** [comment]
> When can we expect to see the 1.6.0 update on the release/stable/main channel/branch? (firefox xpi)

Probably tomorrow

**Comment by Bun01294** [comment]
I get actual ads leaking instead of commerical break in progress in the beta 1.6.0. I don't get any of these in the latest stable build and i have the stable ttv lol pro extension disabled when i'm using the beta build.

**Comment by Bun01294** [comment]
Nvm, i'm getting this in the stable builds.

**Comment by younesaassila** [comment]
> I'm getting actual ads that are leaking instead of commerical break in progress in the beta 1.6.0. I don't get any of these in the latest stable build and i have the stable ttv lol pro extension disabled when i'm using the beta build.

@Bun01294 Were these ads prerolls or midrolls? TTV LOL PRO v1.6.0 only attempts to skip midrolls.

**Comment by Bun01294** [comment]
> > I'm getting actual ads that are leaking instead of commerical break in progress in the beta 1.6.0. I don't get any of these in the latest stable build and i have the stable ttv lol pro extension disabled when i'm using the beta build.
> 
> @Bun01294 Were these ads prerolls or midrolls? TTV LOL PRO v1.6.0 only attempts to skip midrolls.

They are prerolls on version 1.6.0 and the latest stable build 1.5.1

**Comment by younesaassila** [comment]
> Its prerolls on version 1.6.0 and version 1.5.1

Yeah those sometimes get through, unfortunately. I believe this is due to a recent change Twitch made on how they serve ads.

**Comment by Bun01294** [comment]
I see, that's a bummer

**Comment by Bun01294** [comment]
Also is #32 meant for reseting midrolls? i may have jumped the gun a bit too quick, i was getting these ads recently like yesterday

**Comment by juliocoliveira** [comment]
I just got a midroll, Played the stream for 5-10 minutes, then 3 ads was show, Updated the extension by itself today.

**Comment by younesaassila** [comment]
> I just got a midroll, Played the stream for 5-10 minutes, then 3 ads was show, Updated the extension by itself today.

Did the stream reset (quick black screen) and then an ad was shown (preroll)? Or was it just a midroll with no reset whatsoever?

**Comment by juliocoliveira** [comment]
Usually I see the ads for half second and it reset. Today it didn't reset, No quick black screen, Just the ad was show, didn't saw the console to see if the extension detected the ad.

**Comment by juliocoliveira** [comment]
It happened only once, I'm in on twitch now and not seeing any ads.

**Comment by juliocoliveira** [comment]
I just got a reset, it seems to be working fine now. No preroll so far.

**Comment by younesaassila** [comment]
That's unfortunate. Yeah please let me know if you happen to see another ad where the extension doesn't print output to the console. Let's hope the feature catches most midrolls (until Twitch either breaks the Reset player feature or changes how ads are served, again).

**Comment by juliocoliveira** [comment]
I'm getting prerolls now, But if I reload the page it does not show up anymore, can't you @younesaassila use the same script to refresh the player for prerolls too? as it works for midrolls?

**Comment by younesaassila** [comment]
> I'm getting prerolls now, But if I reload the page it does not show up anymore, can't you @younesaassila use the same script to refresh the player for prerolls too? as it works for midrolls?

The issue with prerolls is you might get into an infinite refresh loop. Let's hope the TTV LOL team can figure out how to remove prerolls like they used to do.

**Comment by juliocoliveira** [comment]
I very much doubt that, the last update is 2 years old, TTV LOL extensions is kind of abandoned. But the servers are still running.

**Comment by juliocoliveira** [comment]
A got another midroll without reset. Those are rare but are happening:
![image](https://user-images.githubusercontent.com/6606725/210154280-99794eff-581b-41cc-97db-30cf19ec3b22.png)

**Comment by younesaassila** [comment]
Seems like the injected script isn't always injected successfully. The code would need to check if it's been injected properly. I'll work on that.

In the meantime, you can install FrankerFaceZ. TTV LOL PRO is coded so that if it detects FFZ's Reset player button it'll click it instead of trying to communicate to the injected script.

**Comment by thinkpad4** [comment]
@juliocoliveira Remember that all TTVLOL official extension and this extension at its base do is load the API server.  There is no need for the official TTVLOL extension to be updated because all changes are done server side not extension side.  So them not updating it for 2 years doesn't mean much as long as they are maintaining the API server

**Comment by younesaassila** [comment]
Just released [version 1.6.1](https://github.com/younesaassila/ttv-lol-pro/releases/tag/v1.6.1) that _hopefully_ improves the reliability of the "Reset player on midroll" feature. (Still experimental, still Firefox only)

**Comment by Bun01294** [comment]
still have to reload the page to remove the midrolls, on the other hand i did find the new version working somewhat on reseting the player when it encounters a midroll

**Comment by younesaassila** [comment]
Hey, rewrote the feature to make it work on Chrome. Just published a [beta release](https://github.com/younesaassila/ttv-lol-pro/releases/tag/v1.7.0-beta.0) with this feature. If you encounter any bug on Chrome or Firefox, please report them in this thread.
Thanks :)

**Comment by Void48** [comment]
> Hey, rewrote the feature to make it work on Chrome. Just published a [beta release](https://github.com/younesaassila/ttv-lol-pro/releases/tag/v1.7.0-beta.0) with this feature. If you encounter any bug on Chrome or Firefox, please report them in this thread. Thanks :)

Not working very well.. Player won't load 90% of the time and I have to disable extension then refresh site
![chrome_pgT6OD5zIF](https://user-images.githubusercontent.com/85196642/211129148-c441213e-3f2a-46d1-9a1a-650751b9bbe7.gif)

**Comment by younesaassila** [comment]
@Void48 Could you post a screenshot of the page's console when that happens? That way I can check what errors are happening

**Comment by younesaassila** [comment]
@Void48 Tried to replicate your issue but couldn't. Can you provide more context? Was this happening when trying to initially load a stream or was it happening mid-way (as in "a midroll could have tried to load")? Also what browser are you using?

**Comment by Void48** [comment]
> @Void48 Tried to replicate your issue but couldn't. Can you provide more context? Was this happening when trying to initially load a stream or was it happening mid-way (as in "a midroll could have tried to load")? Also what browser are you using?

Chrome
![chrome_K0YCfjQiMJ](https://user-images.githubusercontent.com/85196642/211166378-f35fbcf7-2dbc-42c2-963d-425e4eb5c987.png)
![chrome_mp3ZmvdIzm](https://user-images.githubusercontent.com/85196642/211166379-c0eee57e-9210-400a-ae80-2d8205f0afe5.png)

**Comment by younesaassila** [comment]
Okay thanks I see the problem. Going to investigate

**Comment by younesaassila** [comment]
I've released a [new beta](https://github.com/younesaassila/ttv-lol-pro/releases/tag/v1.7.0-beta.1) that _hopefully_ fixes the above issue, improves the reliability of the "Reset player on midroll" feature, and add logs.

@Void48 If the issue persists in the new beta, would you mind sharing a screenshot of the page's console? I've added logs to let me better debug the problem if it persists

**Comment by Void48** [comment]
> I've released a [new beta](https://github.com/younesaassila/ttv-lol-pro/releases/tag/v1.7.0-beta.1) that _hopefully_ fixes the above issue, improves the reliability of the "Reset player on midroll" feature, and add logs.
> 
> @Void48 If the issue persists in the new beta, would you mind sharing a screenshot of the page's console? I've added logs to help me better debug the problem if it persists

seems to work better now. no idea about how well the reset player functionality works though

**Comment by jakethewake** [comment]
The new beta seems to work for me with midrolls on Firefox dev edition.
Refreshed 4 or 5 times but no ads after that.
![bild](https://user-images.githubusercontent.com/59709637/211612183-6f43fef6-2284-4d4c-a4d7-78bca24f829b.png)

**Comment by younesaassila** [comment]
@jakethewake Thanks for the feedback, I'll try to add code to prevent multiple resets in a row

**Comment by juliocoliveira** [comment]
I think its best to reset automatically 5 times than manually refresh 5 times.

**Comment by younesaassila** [comment]
@juliocoliveira Of course, but I believe one reset would be enough to skip a midroll (and less resets = better user experience). I'm currently testing if the one-reset solution works on my end

**Comment by juliocoliveira** [comment]
Oh yeah sure. If you manage to do a 5 reset one time is much better.

**Comment by younesaassila** [comment]
Can't seem able to get an ad (it's always when you want them that they don't show up lol) to test if one reset is enough.

Released [a new beta](https://github.com/younesaassila/ttv-lol-pro/releases/tag/v1.7.0-beta.2) with this change if you'd like to test if the feature works.

**Comment by juliocoliveira** [comment]
Tried to install it on the developer edition, but i got this:
![image](https://user-images.githubusercontent.com/6606725/211671403-0674bba2-270b-4533-a26a-da41f182303c.png)
Tried to unzip, and zip it again, restart browser, nothing works.

**Comment by younesaassila** [comment]
@juliocoliveira That's weird. I've rebuilt the ZIP file, could you try to download it and try again?

**Comment by juliocoliveira** [comment]
As temporary i can install, But as normal addon i cannot, I turned off xpinstall.signatures.required on the developer edition, But still says its corrupt.

**Comment by younesaassila** [comment]
@juliocoliveira I didn't even know that was possible 😅
Was this working with the previous beta? Perhaps it's a Firefox bug?

**Comment by juliocoliveira** [comment]
Never mind, I forgot that you have to add: 
```
 "browser_specific_settings": {
        "gecko": {
            "id": "test@gmail.com"
        }
    }```
To the manifest to that feature work, I managed to install it now. I'll test today and see if there is no multiple resets.

**Comment by juliocoliveira** [comment]
What i saw wrong is the volume after a reset. Before the reset the volume is 100% after the reset the volume is like 10%

**Comment by juliocoliveira** [comment]
I had a 3 reset in a row:
![image](https://user-images.githubusercontent.com/6606725/211808598-07f54a6a-9176-47a4-81d5-a34c9248a172.png)

**Comment by younesaassila** [comment]
> I had a 3 reset in a row:
> 
> ![image](https://user-images.githubusercontent.com/6606725/211808598-07f54a6a-9176-47a4-81d5-a34c9248a172.png)
> 
> 

Yeah this is the best that can be done. When resetting the player a new manifest is fetched and sometimes it includes a midroll too. The extension still prevented too many resets as you can see in the logs.

**Comment by juliocoliveira** [comment]
Ahm, OK, see if you can fix the volume, its very annoying.

**Comment by Mioni** [comment]
I'm just here to thank you for this feature, very good!
No problem so far.

**Comment by juliocoliveira** [comment]
Ahm, Thanks!, I tested and the issue about the volume seams gone, but i didn't got multiple resets yet. So too soon to say it for sure. I'll keep testing, In a day or two I say if it really got fixed.

**Comment by younesaassila** [comment]
> Ahm, Thanks!, I tested and the issue about the volume seems gone, but i didn't got multiple resets yet. So too soon to say it for sure. I'll keep testing, In a day or two I say if it really got fixed.

In this beta, you can manually trigger a player reset by typing `window.postMessage({type:"resetPlayer"})` in the page console.

**Comment by juliocoliveira** [comment]
I tested with the command, `window.postMessage({type:"resetPlayer"})` reset it 5 times in a row, and the volume stays at 100%. So you fixed it.

**Comment by juliocoliveira** [comment]
Now what is bothering me is that i some times see an Ad for 10 seconds before the player reset it, Can it be improved and shorter?

**Comment by younesaassila** [comment]
> Now what is bothering me is that i some times see an Ad for 10 seconds before the player reset it, Can it be improved and shorter?

Can you check the console to see if the extension detected the midroll before the ad started (in your example at least 10 seconds before)?

E.g.: The extension logged "Midroll scheduled for XXX (in 10000 ms)"

**Comment by juliocoliveira** [comment]
[TTV LOL PRO] Midroll scheduled for 2023-01-17T16:18:43.755Z (in 18353 ms)
[TTV LOL PRO] Resetting player… 
[TTV LOL PRO] Midroll scheduled for 2023-01-17T16:19:03.608Z (in 17596 ms)

This time it took 5 seconds.

**Comment by juliocoliveira** [comment]
This resets that take 5 or 10 seconds to start is few, Most resets I don't see the ad.

**Comment by juliocoliveira** [comment]
This took 10 seconds:
![image](https://user-images.githubusercontent.com/6606725/212960968-1233b6f9-47e9-4fdb-8069-e4b0b5af84fc.png)

**Comment by younesaassila** [comment]
@juliocoliveira Ok I've removed the scheduling part and now resetting the player immediately. Available in new beta: https://github.com/younesaassila/ttv-lol-pro/releases/tag/v1.7.0-beta.4

Can you tell me if that works better? (I fear this might lead to more resets in a row)

**Comment by juliocoliveira** [comment]
Seems to be working just fine, No problems so far, Not seeing any Ads so far, I think this is ready to be released.

**Comment by younesaassila** [comment]
> Seems to be working just fine, No problems, Not seeing any Ads so far, I think this is ready to be released.

Great news! Though I'm worried the feature might break things for users who combine TTV LOL PRO with the Vaft script (since both check the Video Weaver response to determine if an ad is about to play). I'm thinking of disabling the feature by default, perhaps.

**Comment by juliocoliveira** [comment]
What is exactly the feature you planning to disable? Player reset? People should NOT use multiple specific ad blockers, it's just not a good practice.
I think you should not disable any feature by default

**Comment by younesaassila** [comment]
🎉 Now available for Chrome and Firefox in version 1.7.0.
I've removed unused parts of the code, let's hope for a smooth rollout. If any of you encounter a problem, please create a new issue so it can get fixed asap.

**Comment by younesaassila** [comment]
Just a quick note: The feature is now disabled by default starting in v1.7.1 due to incompatibilities with the VAFT script.
